linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <semaphore.h>
+#include <time.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/gballoc_hl.h" // IWYU pragma: keep
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS          100
+#define MAX_THREAD_COUNT    1
+#define THREAD_LIMIT        65536
+
+#define TASK_RESULT_VALUES  \
+    TASK_NOT_USED,          \
+    TASK_WAITING,           \
+    TASK_WORKING
+
+MU_DEFINE_ENUM(TASK_RESULT, TASK_RESULT_VALUES);
+MU_DEFINE_ENUM_STRINGS(TASK_RESULT, TASK_RESULT_VALUES)
+
+#define THREADPOOL_STATE_VALUES \
+    THREADPOOL_STATE_NOT_OPEN, \
+    THREADPOOL_STATE_OPEN
+
+MU_DEFINE_ENUM(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+MU_DEFINE_ENUM_STRINGS(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    volatile_atomic int32_t task_state;
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Array
+    THREADPOOL_TASK* task_array;
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+
+    THREAD_HANDLE* thread_handle_list;
```

Comment:
>THREAD_HANDLE* [](http://example.com/codeflow?start=4&length=14)

Consider using flexible arrays for this. One less allocation.
Consider as it is not expected to happen very often anyway.

Reply:
Will be a todo later